packages/frontend/src/hooks/useKeycard.ts

```diff
@@ -1,27 +1,83 @@
+import { getLogger } from "@logtape/logtape";
+import { useQuery, useQueryClient } from "@tanstack/react-query";
+
+import { RelayClient } from "@massmarket/client";
+import { getWindowLocation } from "@massmarket/utils";
 
 import { useShopId } from "./useShopId.ts";
 import { KeycardRole } from "../types.ts";
+import { useRelayEndpoint } from "./useRelayEndpoint.ts";
+
+const logger = getLogger(["mass-market", "frontend", "useKeycard"]);
 
-export function useKeycard(
-  keycard?: {
-    privateKey: `0x${string}`;
-    role: KeycardRole;
-    address: `0x${string}`;
-  },
-) {
-  if (!keycard) {
-    const privateKey = generatePrivateKey();
-    const account = privateKeyToAccount(privateKey);
-    keycard = {
-      privateKey,
-      role: KeycardRole.NEW_GUEST,
-      address: account.address,
-    };
-  }
+/**
+ * This hook returns a keycard for the browser session.
+ * The async queryFn where we enroll the keycard only executes once unless variables in the queryKey changes.
+ * The keycard will be cached for the duration of the browser session regardless of refreshes.
+ */
+export function useKeycard(role: KeycardRole = "guest") {
   const { shopId } = useShopId();
-  const keyCardID = "keycard" + shopId;
-  // console.log({shopId, keyCardID});
-  return useLocalStorage(keyCardID, keycard);
+  const { data: wallet } = useWalletClient();
+  const { address } = useAccount();
+  const { relayEndpoint } = useRelayEndpoint();
+  const queryClient = useQueryClient();
+
+  return useQuery({
+    // queryFn will not execute till these variables are defined.
+    enabled: !!shopId && !!wallet && !!relayEndpoint && !!address,
+    queryKey: [
+      "keycard",
+      address,
+      role,
+    ],
+    queryFn: async () => {
+      /**
+       * 1. Generate KC
+       * 2. Enroll KC
+       * 3. Return the KC
+       */
+
+      const privateKey = generatePrivateKey();
+
+      // This relay instance is just to enroll the keycard.
+      const relayClient = new RelayClient({
+        relayEndpoint,
+        walletClient: wallet,
+        keycard: privateKeyToAccount(privateKey),
+        shopId,
+      });
+
+      const res = await relayClient.enrollKeycard(
+        wallet,
+        address,
+        role === "guest",
+        getWindowLocation(),
+      );
+
+      if (!res.ok) {
+        throw new Error(`Failed to enroll keycard: ${res.status}`);
+      }
+      logger.debug(`Success: Enrolled new ${role} keycard`);
+
+      const kc = {
+        privateKey,
+        role,
+        address,
+      };
+      // Return this keycard for all guest keycard queries.
+      // This is needed for merchant enrolls, since we don't want enrolled merchant keycards to be overwritten by guest keycards.
+      // setQueriesData also ensures that any component using the query will re-render with the new keycard.
+      queryClient.setQueriesData(
+        { queryKey: ["keycard", address, "guest"] },
+        kc,
+      );
```

Comment:
we only want to set this when enrolling a merchant. Also in the comment its say " since we don't want enrolled merchant keycards to be overwritten by guest keycards." this is not quite what we are doing, we are overwritting the the guest keycard with the merchants 
```suggestion

if (merchant) {
      queryClient.setQueriesData(
        { queryKey: ["keycard", address, "guest"] },
        kc,
      );
    }
```